src/components/heatmap/heatmap.tsx

```diff
@@ -0,0 +1,344 @@
+    if (invalidConfigs.length > 0) {
+      setInvalidConfigs(() => invalidConfigs);
+      setConfigErrorsPresent(true);
+    }
+  }
+
+  /**
+   * @param page : Page object fecthed from response of ermrestURI
+   */
+  const readAll = (page: any) => {
+    for (let i = 0; i < page.tuples.length; i++) {
+      addData(page.tuples[i]);
+    }
+    if (page.hasNext) {
+      Object(header).action = 'page';
+      page.next.read(1000, header).then(readAll).catch(function () {
+        setInvalidConfigs(() => ['Error while reading data from Ermrestjs']);
+        setConfigErrorsPresent(() => false);
+      });
+    } else {
+      setHeatmaps(() => (heatmaps));
+    }
+    setNCBIGeneID(() => (page.tuples[0].data.NCBI_GeneID));
+  }
+  /**
+   * 
+   * @param tuple : Tuple object retrieved from page
+   * Adds data(x,y,z values) to heatmap objects
+   */
+
+  const addData = (tuple: any) => {
+    const configData = Object.assign({}, config.data);
+    let hm = null;
+    const x = tuple.data[configData.xColumn];
+    const y = tuple.data[configData.yColumn];
+    const z = tuple.data[configData.zColumn];
+    const title = tuple.data[configData.titleColumn];
+    const id = tuple.data[configData.idColumn];
+    for (let i = 0; i < heatmaps.length; i++) {
+      if (Object(heatmaps[i]).title === title) {
+        Object(heatmaps)[i]['id'] = id;
+        hm = heatmaps[i];
+      }
+    }
+    if (hm === null) {
+      hm = { 'title': title, 'rows': { y: [], x: [], z: [], type: 'heatmap' } };
+      heatmaps.push(hm);
+    }
+    let rowIndex = Object(hm).rows.y.indexOf(y);
+    if (rowIndex < 0) {
+      Object(hm).rows.y.push(y);
+      Object(hm).rows.z.push([]);
+      rowIndex = Object(hm).rows.y.indexOf(y);
+    }
+    if (!Object(hm).rows.x.includes(x)) {
+      Object(hm).rows.x.push(x);
+    }
+    Object(hm).rows.z[rowIndex].push(z);
+  }
+
+  /**
+     * @param {object} input : Input parameters of heatmap directive
+     * @param {number} longestXTick : Length of longest X axis label
+     * @param {number} longestYTick : Length of longest Y axis label
+     * @param {number} lengthY : Number of Y values
+     * Calculates the height and margins of the heatmap based on the number of y values and length of the longest X label
+     * so that the labels do not get clipped and the bar height is adjusted accordingly.
+     * Return an object with all the required layout parameters.
+     * @example
+     * {
+     * 	height: height of the heatmap,
+     * 	width: width of the heatmap,
+     * 	margin: {
+     * 		t: top margin of the heatmap,
+     * 		r: right margin of the heatmap,
+     * 		b: bottom margin of the heatmap,
+     * 		l: left of the heatmap
+     * 	},
+     * 	xTickAngle: inclination of x axis labels,
+     *  yTickAngle: inclination of y axis labels,
+     * 	tickFont: font to be used in labels
+     * }
+     */
+  const getHeatmapLayoutParams = (input: any, longestXTick: number, longestYTick: number, lengthY: number) => {
```

Comment:
`input` should have a type defined (see comment below).

The function should also have a return type defined since we are intending for this to always be called one way and always return a `layoutParams` object we are defining ourself.